### ISSUE TITLE
Object literal expanded to the wrong union type arm

### ISSUE THREAD

**Issue by frankpf** [Microsoft/TypeScript#22019]
**TypeScript Version:**  2.8.0-dev.20180217

**Search Terms:** "is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes", "Property '[Symbol.unscopables]' is missing in type"
                       

**Code**

bug.tsx:
```tsx
// Test this by running tsc --jsx
import * as React from 'react';
import { View } from 'react-native'

const screen = {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
}

export function Login() {
  return(
    <View style={ screen }>
    </View>
  )
}
```

I also created a repository with the correct dependencies to make debugging this easier:
https://github.com/frankpf/tsc-react-native-bug

**Expected behavior:**
No error.

**Actual behavior:**
On the <View style={ screen }>, I get this error:

```
src/bug.tsx(12,11): error TS2322: Type '{ children: never[]; style: { flex: number; justifyContent: string; alignItems: string; }; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<ViewProperties, ComponentState>> & Reado...'.
  Type '{ children: never[]; style: { flex: number; justifyContent: string; alignItems: string; }; }' is not assignable to type 'Readonly<ViewProperties>'.
    Types of property 'style' are incompatible.
      Type '{ flex: number; justifyContent: string; alignItems: string; }' is not assignable to type 'StyleProp<ViewStyle>'.
        Type '{ flex: number; justifyContent: string; alignItems: string; }' is not assignable to type 'RecursiveArray<false | ViewStyle | RegisteredStyle<ViewStyle> | null | undefined>'.
          Property '[Symbol.unscopables]' is missing in type '{ flex: number; justifyContent: string; alignItems: string; }'.
```

One super weird thing is that **if I just pass the object literal, I get no error**. So
`<View style={ screen }>`
errors, but this 
`<View style= {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center'
 }}>`
does not.

If I remove the `justifyContent` and `alignItems` properties from the `screen` variable:
```tsx
const screen = {
    flex: 1
}
```

tsc also doesn't complain.

**Related Issues:** https://github.com/Microsoft/TypeScript/issues/15463 and https://github.com/Microsoft/TypeScript/issues/15419 may be slightly related

**Comment by frankpf** [comment]
I reduced this issue a lot. 

```typescript
import { StyleProp, ViewStyle, RecursiveArray, RegisteredStyle } from 'react-native'

const screen: ViewStyle = {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
}

let x: StyleProp<ViewStyle> = screen
```
This compiles with no error, because `screen` is a valid `ViewStyle`.

However, if I simply remove the type hint from `screen`:
```typescript
import { StyleProp, ViewStyle } from 'react-native'

const screen = {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
}

let x: StyleProp<ViewStyle> = screen
```

then tsc outputs the following on the `let x` assignment:
```typescript
src/bug.tsx(9,5): error TS2322: Type '{ flex: number; justifyContent: string; alignItems: string; }' is not assignable to type 'StyleProp<ViewStyle>'.
  Type '{ flex: number; justifyContent: string; alignItems: string; }' is not assignable to type 'RecursiveArray<false | ViewStyle | RegisteredStyle<ViewStyle> | null | undefined>'.
    Property '[Symbol.unscopables]' is missing in type '{ flex: number; justifyContent: string; alignItems: string; }'.
```

Here are the definitions of `StyleProp<T>` and RecursiveArray<T>, taken from `@types/react-native`: 
```typescript
interface RecursiveArray<T> extends Array<T | RecursiveArray<T>> {}
export type StyleProp<T> = T | RegisteredStyle<T> | RecursiveArray<T | RegisteredStyle<T> | Falsy> | Falsy```;
```

To me, it seems that **the main issue is that `tsc` is wrongly expanding `screen` to the `RecursiveArray<T | ...>` part of the `StyleProp` union instead of expanding it to just `T`**. `screen` is a valid `ViewStyle`, but it's not a `RecursiveArray`.

**Comment by Rovanion** [comment]
What resolved this issue?

**Comment by frankpf** [comment]
Sorry, I should've explained the solution before closing the issue.

**TLDR**:

Use React Native's `StyleSheet.create`:
```tsx
import { StyleSheet } from 'react-native'
const screen = StyleSheet.create({
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
})```

**Long version**:

```typescript
const x = 'literal'

const y = {
   x: 'literal'
}
```

If you check the types, `x` is of type `'literal'` but `y.x` is a `string`. This makes sense because `y.x` is mutable and `x` is not.

The unintuitive part happens when you assign y to an interface expecting a literal string type:
```typescript
interface A {
   x: 'literal'
}
const y = {
   x: 'literal'
}

let a: A = y <-- tsc error: "Type string is not assignable to type literal"
```

OK, this is a bit unexpected but it's understandable. The problem happens when you use a union type:

```typescript
interface A {
   x: 'literal'
}
const y = {
   x: 'literal'
}

type B = A | number[]

let a: B = y <-- tsc error: "Property 'length' is missing in type { x: string }"
```

This is especially problematic for big libraries, as you get very long, non intuitive error message (as you can see in the OP).

It seems that this only happens for some union types.

`type B = A | number`, for instance, shows the "correct" error message. I thought this only happened when arrays were in the union types, but `interface Z { z: 'abc' }; type B = A | Z` also errors with `Property "z" is missing in { x: string }`.
